File: oracles/q_closure_predictive_v147.exs
defmodule QClosurePredictiveV147 do
  def positive(n), do: n

  def weighted(h, x), do: h * x

  def cap(0, _value), do: 0
  def cap(_budget, 0), do: 0
  def cap(budget, value), do: 1 + cap(budget - 1, value - 1)

  def project(budget, h, x), do: cap(budget, weighted(h, x))

  def oracle_identity(n), do: positive(n) == n
end

true = QClosurePredictiveV147.oracle_identity(7)

File: oracles/QClosurePredictive_v147.ml
let positive n = n

let weighted h x = h * x

let rec cap budget value =
  match budget, value with
  | 0, _ -> 0
  | _, 0 -> 0
  | b, v -> 1 + cap (b - 1) (v - 1)

let project budget h x = cap budget (weighted h x)

let oracle_identity n = positive n = n

let () =
  assert (oracle_identity 7)
